### PR TITLE
added dark mode to map and refatoring done

### PR DESCRIPTION
**Pull request by isham-mohamed**

### CODE REVIEW

File: app/src/main/java/com/example/androidDeviceDetails/collectors/LocationCollector.kt [hunk 1]
```diff
@@ -9,8 +9,8 @@ import android.location.LocationManager.GPS_PROVIDER
 import android.location.LocationManager.NETWORK_PROVIDER
 import android.util.Log
 import com.example.androidDeviceDetails.base.BaseCollector
+import com.example.androidDeviceDetails.models.database.LocationModel
 import com.example.androidDeviceDetails.models.database.RoomDB
-import com.example.androidDeviceDetails.models.location.LocationModel
 import com.example.androidDeviceDetails.utils.Utils
 import kotlinx.coroutines.GlobalScope
 import kotlinx.coroutines.launch
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/LocationCollector.kt:19]: - add one line permission check insted of suppressing 
- hasGps and hasNetwork can be made local to the start function
- the first if condition with the OR can be avoided
- locationNetwork and locationGps can be made lateinint and do : :isInitialized check and avoid all null "!!"

File: app/src/main/java/com/example/androidDeviceDetails/models/location/LocationItemViewHolder.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.example.androidDeviceDetails.models.location
+
+import com.example.androidDeviceDetails.adapters.LocationAdapter
+
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/models/location/LocationItemViewHolder.kt]: - This is just a data class not view holder rename accordingly.

File: app/src/main/java/com/example/androidDeviceDetails/adapters/LocationAdapter.kt [hunk 3]
```diff
@@ -7,11 +7,11 @@ import android.widget.TextView
 import androidx.recyclerview.widget.RecyclerView
 import com.example.androidDeviceDetails.R
 import com.example.androidDeviceDetails.interfaces.OnItemClickListener
-import com.example.androidDeviceDetails.models.location.LocationDisplayModel
+import com.example.androidDeviceDetails.models.location.LocationItemViewHolder
 import com.example.androidDeviceDetails.utils.SortBy
 
 
-class LocationAdapter(var dataSet: ArrayList<LocationDisplayModel>, private val onItemClickListener: OnItemClickListener):
+class LocationAdapter(var dataSet: ArrayList<LocationItemViewHolder>, private val onItemClickListener: OnItemClickListener):
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/LocationAdapter.kt]: - dataSet can be made private

File: app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt [hunk 4]
```diff
@@ -19,34 +18,31 @@ class LocationCooker : BaseCooker() {
     private val geoCoder: Geocoder = Geocoder(DeviceDetailsApplication.instance)
 
 
-    private fun cookData(locationList: ArrayList<LocationModel>): ArrayList<LocationDisplayModel> {
+    private fun cookData(locationList: ArrayList<LocationModel>): ArrayList<LocationItemViewHolder> {
         val cookedLocationList = emptyList<String>().toMutableList()
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt]: - cookedLocationList  = ArrayList<String>

File: app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt [hunk 5]
```diff
@@ -19,34 +18,31 @@ class LocationCooker : BaseCooker() {
...
         var prevLocationHash = ""
-        for (location in locationList) {
-            val newHash =
-                GeoHash.encodeHash(location.latitude!!, location.longitude!!, geoHashLength).toString()
+        for (loc in locationList) {
+            val newHash = GeoHash.encodeHash(loc.latitude, loc.longitude, geoHashLength).toString()
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt]: - .toString() is ont required .encodeHash returns string

File: app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt [hunk 6]
```diff
@@ -19,34 +18,31 @@ class LocationCooker : BaseCooker() {
...
             if (newHash != prevLocationHash) {
                 prevLocationHash = newHash
                 cookedLocationList.add(newHash)
             }
         }
         val countedData = cookedLocationList.groupingBy { it }.eachCount()
-        val locationDisplayModel: ArrayList<LocationDisplayModel> = ArrayList()
+        val locationItemViewHolder: ArrayList<LocationItemViewHolder> = ArrayList()
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt]: - directly initialise =  ArrayList<LocationItemViewHolder>( )
- this is a list so renave accordingly

File: app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt [hunk 7]
```diff
@@ -19,34 +18,31 @@ class LocationCooker : BaseCooker() {
...
         for ((k,v) in countedData){
             val latLong = GeoHash.decodeHash(k)
             val address = geoCoder.getFromLocation(latLong.lat, latLong.lon, 1)[0]?.locality?.toString()
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt]: - .toString() is ont required

File: app/src/main/java/com/example/androidDeviceDetails/viewModel/location/LocationViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        binding.root.post {
+            val mapController = binding.mapView.controller
+            mapController.setZoom(15.0)
+            mapController.setCenter(geoPoint)
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T> onDone(outputList: ArrayList<T>) {
+        cookedDataList = outputList as ArrayList<LocationItemViewHolder>
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/viewModel/location/LocationViewModel.kt]: - avoid @Suppress("UNCHECKED_CAST") by using .filterIsInstance<LocationItemViewHolder>() as ArrayList wherever possible

File: app/src/main/java/com/example/androidDeviceDetails/viewModel/location/LocationViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            drawable.intrinsicWidth, drawable.intrinsicHeight, Bitmap.Config.ARGB_8888
+        )
+        val canvas = Canvas(bitmap)
+        drawable.setBounds(0, 0, canvas.width, canvas.height)
+        drawable.draw(canvas)
+        return bitmap
+    }
+
+    private fun addPointOnMap() {
+        val overlays: MutableList<Overlay> = binding.mapView.overlays
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/viewModel/location/LocationViewModel.kt]: -  : MutableList<Overlay> is not required you can directly assign

File: app/src/main/java/com/example/androidDeviceDetails/ui/LocationActivity.kt [hunk 10]
```diff
@@ -10,21 +14,22 @@ import com.example.androidDeviceDetails.adapters.LocationAdapter
...
 import java.util.*
 import kotlin.collections.ArrayList
+import org.osmdroid.config.Configuration as osmConfig
 
 
-class LocationActivity : AppCompatActivity(), View.OnClickListener,OnItemClickListener {
-    private lateinit var activityController: ActivityController<LocationDisplayModel>
+class LocationActivity : AppCompatActivity(), View.OnClickListener, OnItemClickListener {
+    private lateinit var activityController: ActivityController<LocationItemViewHolder>
     lateinit var locationViewModel: LocationViewModel
```
Comment (richardjoy530) [inline app/src/main/java/com/example/androidDeviceDetails/ui/LocationActivity.kt]: - can make locationViewModel private

**Review by richardjoy530 — CHANGES_REQUESTED** [review]
- Review over

File: app/src/main/java/com/example/androidDeviceDetails/models/location/LocationItemViewHolder.kt (hunk 2 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/models/location/LocationItemViewHolder.kt]: done

File: app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt (hunk 4 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt]: done

File: app/src/main/java/com/example/androidDeviceDetails/adapters/LocationAdapter.kt (hunk 3 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/adapters/LocationAdapter.kt]: done

File: app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt (hunk 5 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt]: done

File: app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt (hunk 6 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt]: done

File: app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt (hunk 7 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/cooker/LocationCooker.kt]: changed

File: app/src/main/java/com/example/androidDeviceDetails/ui/LocationActivity.kt (hunk 10 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/ui/LocationActivity.kt]: done

File: app/src/main/java/com/example/androidDeviceDetails/viewModel/location/LocationViewModel.kt (hunk 9 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/viewModel/location/LocationViewModel.kt]: done

File: app/src/main/java/com/example/androidDeviceDetails/viewModel/location/LocationViewModel.kt (hunk 8 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/viewModel/location/LocationViewModel.kt]: done

File: app/src/main/java/com/example/androidDeviceDetails/collectors/LocationCollector.kt (hunk 1 above)
Comment (isham-mohamed) [reply app/src/main/java/com/example/androidDeviceDetails/collectors/LocationCollector.kt:19]: done

**Review by richardjoy530 — APPROVED** [review]

**Pull request merged**